File: src/libs/posix/src/sys/uio/syscall/preadv.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use crate::{
    limits,
    sys::{
        types::{
            off_t,
            size_t,
            ssize_t,
        },
        uio::iovec,
    },
    unistd,
};
use ::nvx::sys::error::ErrorCode;

//==================================================================================================
// Standalone Functions
//==================================================================================================

///
/// # Description
///
/// Reads a vector of data from a file.
pub fn preadv(fd: i32, iov: *const iovec, iovcnt: i32, offset: off_t) -> ssize_t {
    // Check if number of elements in the vector is valid.
    if (iovcnt < 0) || (iovcnt > limits::IOV_MAX as i32) {
        return (-ErrorCode::InvalidArgument.into_errno()) as ssize_t;
    }

    // Check if vector base is invalid.
    if iov.is_null() {
        return (-ErrorCode::InvalidArgument.into_errno()) as ssize_t;
    }

    // Check for zero-length vector.
    if iovcnt == 0 {
        return 0;
    }

    // Check if offset is invalid.
    if offset < 0 {
        return (-ErrorCode::InvalidArgument.into_errno()) as ssize_t;
    }

    let mut total: ssize_t = 0;

    // Traverse i/o vector.
    for i in 0..iovcnt {
        let iov: *const iovec = unsafe { iov.offset(i as isize) };

        // Check if base address is invalid.
        if iov.is_null() {
            return (-ErrorCode::InvalidArgument.into_errno()) as ssize_t;
        }

        let iov_base: *mut u8 = unsafe { (*iov).iov_base };
        let iov_len: size_t = unsafe { (*iov).iov_len };

        // Check if base address is invalid.
        if iov_base.is_null() {
            return (-ErrorCode::InvalidArgument.into_errno()) as ssize_t;
        }

        // Read data.
        let count: ssize_t = unistd::pread(fd, iov_base, iov_len, offset + total as off_t);

        // Check if read failed.
        if count < 0 {
            return count;
        }

        total += count;
    }

    total
}

File: src/tests/file-c/preadv.c
/*
 * Copyright(c) The Maintainers of Nanvix.
 * Licensed under the MIT License.
 */

//==================================================================================================
// Configuration
//==================================================================================================

/* Must come first. */
#define _BSD_SOURCE // preadv()

//==================================================================================================
// Imports
//==================================================================================================

#include <assert.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>
#include <sys/uio.h>
#include <unistd.h>

//==================================================================================================
// Constants
//==================================================================================================

// Maximum length for file content.
#define DATA_LEN_MAX 256

//==================================================================================================
// Standalone Functions
//==================================================================================================

// Tests whether we can read from a file using preadv.
void test_preadv(void)
{
    fprintf(stderr, "testing preadv() ... ");

    const char *filename = "testfile.tmp";
    assert(strlen(filename) <= NAME_MAX);

    // Data to write to the file.
    const char *data = "Hello Nanvix!";
    size_t data_len = strlen(data);
    assert(data_len <= DATA_LEN_MAX);

    // Define a constant for the write offset.
    const off_t DATA_OFFSET = 10;

    // Create and open a test file.
    int fd = open(filename, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
    assert(fd != -1);

    // Move the file offset to the specified position.
    assert(lseek(fd, DATA_OFFSET, SEEK_SET) == DATA_OFFSET);

    // Write data to the file at the current offset.
    assert(write(fd, data, data_len) == (ssize_t)data_len);

    // Close the file.
    assert(close(fd) == 0);

    // Reopen the file for reading.
    fd = open(filename, O_RDONLY);
    assert(fd != -1);

    // Prepare buffers for preadv.
    char buffer1[7]; // "Hello "
    char buffer2[8]; // "Nanvix!"
    struct iovec iov[2];
    iov[0].iov_base = buffer1;
    iov[0].iov_len = 6; // Length of "Hello "
    iov[1].iov_base = buffer2;
    iov[1].iov_len = 7; // Length of "Nanvix!"

    // Read data using preadv at the specified offset.
    ssize_t bytes_read = preadv(fd, iov, 2, DATA_OFFSET);
    assert(bytes_read == (ssize_t)data_len);

    // Null-terminate the buffers.
    buffer1[6] = '\0';
    buffer2[7] = '\0';

    // Assert contents.
    assert(strcmp(buffer1, "Hello ") == 0);
    assert(strcmp(buffer2, "Nanvix!") == 0);

    // Close the file.
    assert(close(fd) == 0);

    // Remove the test file.
    assert(unlink(filename) == 0);

    fprintf(stderr, "passed\n");
}
